pkg/helpers/helpers.go

```diff
@@ -163,6 +164,247 @@ func ConcatSlices[T any](slices ...[]T) []T {
+}
+
+type AzureStorageLocation struct {
+	Name                    string   `json:"NAME"`
+	StorageProvider         string   `json:"STORAGE_PROVIDER"`
+	StorageBaseUrl          string   `json:"STORAGE_BASE_URL"`
+	StorageAllowedLocations []string `json:"-"`
+	AzureTenantId           string   `json:"AZURE_TENANT_ID"`
+	AzureMultiTenantAppName string   `json:"-"`
+	AzureConsentUrl         string   `json:"-"`
+	EncryptionType          string   `json:"-"`
+	EncryptionKmsId         string   `json:"-"`
+}
+
+type StorageLocation struct {
+	Name                 string `json:"NAME"`
+	StorageProvider      string `json:"STORAGE_PROVIDER"`
+	StorageBaseUrl       string `json:"STORAGE_BASE_URL"`
+	StorageAwsRoleArn    string `json:"STORAGE_AWS_ROLE_ARN,omitempty"`
+	StorageAwsExternalId string `json:"STORAGE_AWS_EXTERNAL_ID,omitempty"`
+	EncryptionType       string `json:"ENCRYPTION_TYPE,omitempty"`
+	EncryptionKmsKeyId   string `json:"ENCRYPTION_KMS_KEY_ID,omitempty"`
+	AzureTenantId        string `json:"AZURE_TENANT_ID,omitempty"`
+}
+
+func storageLocationsEqual(s1 StorageLocation, s2 StorageLocation) bool {
+	return s1.Name == s2.Name &&
+		s1.StorageProvider == s2.StorageProvider &&
+		s1.StorageBaseUrl == s2.StorageBaseUrl &&
+		s1.StorageAwsRoleArn == s2.StorageAwsRoleArn &&
+		s1.StorageAwsExternalId == s2.StorageAwsExternalId &&
+		s1.EncryptionType == s2.EncryptionType &&
+		s1.EncryptionKmsKeyId == s2.EncryptionKmsKeyId &&
+		s1.AzureTenantId == s2.AzureTenantId
+}
+
+func validateParsedExternalVolumeDescribed(p ParsedExternalVolumeDescribed) error {
+	if len(p.StorageLocations) == 0 {
+		return fmt.Errorf("No storage locations could be parsed from the external volume.")
+	}
+	if len(p.AllowWrites) == 0 {
+		return fmt.Errorf("The external volume AllowWrites property could not be parsed.")
+	}
+
+	for _, s := range p.StorageLocations {
+		if len(s.Name) == 0 {
+			return fmt.Errorf("A storage location's Name in this volume could not be parsed.")
+		}
+		if len(s.StorageProvider) == 0 {
+			return fmt.Errorf("A storage location's StorageProvider in this volume could not be parsed.")
+		}
+		if len(s.StorageBaseUrl) == 0 {
+			return fmt.Errorf("A storage location's StorageBaseUrl in this volume could not be parsed.")
+		}
+
+		storageProvider, err := sdk.ToStorageProvider(s.StorageProvider)
+		if err != nil {
+			return err
+		}
+
+		switch storageProvider {
+		case sdk.StorageProviderS3, sdk.StorageProviderS3GOV:
+			if len(s.StorageAwsRoleArn) == 0 {
+				return fmt.Errorf("An S3 storage location's StorageAwsRoleArn in this volume could not be parsed.")
+			}
+		case sdk.StorageProviderAzure:
+			if len(s.AzureTenantId) == 0 {
+				return fmt.Errorf("An Azure storage location's AzureTenantId in this volume could not be parsed.")
+			}
+		}
+	}
+
+	return nil
+}
+
+type ParsedExternalVolumeDescribed struct {
+	StorageLocations []StorageLocation
+	Active           string
+	Comment          string
+	AllowWrites      string
```

Comment:
Agree that would be cleaner, and it does return "true" so we could change it.

Only consideration is the string type is used to validate that it was set right now - https://github.com/Snowflake-Labs/terraform-provider-snowflake/blob/90f2c88bee1c2388860fd763b217dfe1dbf7fa4b/pkg/helpers/helpers.go#L183. I wouldn't want to change the type to `*bool`, as not having it set is not a valid state. Options I see -
1. Change the type of `AllowWrites ` to `bool`, and run a validation instead that the `ALLOW_WRITES` property is given in https://github.com/Snowflake-Labs/terraform-provider-snowflake/blob/90f2c88bee1c2388860fd763b217dfe1dbf7fa4b/pkg/helpers/helpers.go#L225. Only thing I don't like about this is that the validation isn't happening in the validation function.
2. Duplicate the `ParsedExternalVolumeDescribed`, except changing the type of `AllowWrites ` to `bool`. Before returning, once validation is done, translate the string to a boolean and leave all other fields the same.

Any preferences from your side?

Reply:
As I checked, the field seems to be pretty stable to the outputs it gives. It's either true or false (no empty values, since there is a default for this field). Personally, I would go with the first approach, and in the switch for properties in `ParseExternalVolumeDescribed`, I would assign `strings.ToLower(value) == "true"` to the boolean value (and remove the validation).

Reply:
That’s an interesting point on just deleting the validation. You’re right, it’s not possible on Snowflake that it’s not provided. That’s the case for everything I’m validating in the validation function. So if we’re happy to assume that the SDK is returning the properties as expected we can actually just remove that whole validation function. What do you think?

Reply:
If something changes, then it should be included in the BCR changes bundle (e.g. [this one](https://docs.snowflake.com/en/release-notes/bcr-bundles/2024_08_bundle)). We had previously a few discussions on validations, and we decided to provide minimum validation and let some of the validation happen on the Snowflake side so it won't be mimicked on our side. We decided to still leave schema validation, some of the SDK validation, enum validation, and probably some other validations, but in some cases, we try to do best-effort mappings, so map as much as we can and leave the rest blank. Let's leave it as it is right now, but also leave the thread open (so the person going through the code and open threads will decide what to do).

Reply:
Sounds good 🙂